virtualizarr/readers/zarr.py

```diff
@@ -0,0 +1,372 @@
+        # filepath = 'tmp_2_chunk.zarr'
+
+        # check that Zarr is V3
+        # 1a
+        import zarr
+        from packaging import version
+
+        if version.parse(zarr.__version__).major < 3:
+            raise ImportError("Zarr V3 is required")
+
+        # check_for_collisions will convert them to an empty list
+        drop_variables, loadable_variables = check_for_collisions(
+            drop_variables,
+            loadable_variables,
+        )
+
+        # can we avoid fsspec here if we are using zarr-python for all the reading?
+        # fs = _FsspecFSFromFilepath(filepath=filepath, reader_options=reader_options)
+        ######### tmp############
+
+        # store = zarr.storage.LocalStore(filepath)
+        # zg = zarr.open_consolidated(filepath)
+
+        # 1b.
+
+        zg = zarr.open_group(filepath, mode="r")
+
+        # 2a. Use zarr-python to list the variables in the store
+        zarr_arrays = [val for val in zg.keys()]
+
+        # 2b. and check that all loadable_variables are present
+        assert set(
+            loadable_variables
+        ).issubset(
+            set(zarr_arrays)
+        ), f"loadable_variables ({loadable_variables}) is not a subset of variables in existing Zarr store. This zarr contains:  {zarr_arrays}"
+
+        # virtual variables are available variables minus drop variables & loadable variables
+        virtual_variables = list(
+            set(zarr_arrays) - set(loadable_variables) - set(drop_variables)
+        )
+
+        array_variable_list = []
+        all_array_dims = []
+
+        # 3. For each virtual variable:
+        for var in virtual_variables:
+            # 3a.  Use zarr-python to get the attributes and the dimension names,
+            # and coordinate names (which come from the .zmetadata or zarr.json)
+            array_metadata = zg[var].metadata
+
+            array_metadata_dict = array_metadata.to_dict()
+
+            # extract _ARRAY_DIMENSIONS and remove it from attrs
+            array_dims = array_metadata_dict.get("attributes").pop("_ARRAY_DIMENSIONS")
+
+            # should these have defaults?
+            array_encoding = {
+                "chunks": array_metadata_dict.get("chunks", None),
+                "compressor": array_metadata_dict.get("compressor", None),
+                "dtype": array_metadata_dict.get("dtype", None),
+                "fill_value": array_metadata_dict.get("fill_value", None),
+                "order": array_metadata_dict.get("order", None),
+            }
+
+            # 3b.
+            # Use zarr-python to also get the dtype and chunk grid info + everything else needed to create the virtualizarr.zarr.ZArray object (eventually we can skip this step and use a zarr-python array metadata class directly instead of virtualizarr.zarr.ZArray
+            array_zarray = ZArray(
+                shape=array_metadata_dict.get("shape", None),
+                chunks=array_metadata_dict.get("chunks", None),
```

Comment:
seems like there should be a way to dry this up with the `array_encoding` dict above.